File: AEDSII/TP02/q04/q04.java
import java.util.*;

class q04{
    public static int comp = 0;
    public static int n = 0;
    public static int[] ids = new int[80];
    public static Arq arq = new Arq();
    public static MyIO myIo = new MyIO();

    public static String readArq(int x){
        String str = "";
        for (int i = 0; i < x + 1; i++){
            comp++;
            arq.readLine();
        }   
        str = arq.readLine();
        return str;
    }

    public static int getId(String nome){
        String str = "";
        int id = 0;
        arq.openRead("players.csv");
        //myIo.println("ok");
        do{
            
            if(!str.equals(nome)){
                if (nome.equals("Sarunas Marciulionis")){
                    nome = "Sarunas Marciulionis*";
                }
                str = arq.readLine();
                String[] parts = str.split(",");
                str = parts[1];
                
                
                if (!parts[0].equalsIgnoreCase("id")){
                    id = Integer.parseInt(parts[0]);  
                }
            } 
            
        } while (!str.equalsIgnoreCase(nome));
        //myIo.println(str);
        //myIo.println(str + " " + nome);
        arq.close();
        return id;

    }

    public static boolean pesquisar(String nome){
        boolean resp = false;
        String str = "";
        int dir = n, esq = 0, meio;

        while (esq <= dir){
            arq.openRead("players.csv");
            meio = (esq + dir) / 2;
            
            str = readArq(ids[meio]);
            String[] parts = str.split(",");
            //myIo.println(dir + " " + esq + " " + meio + " " + Integer.parseInt(parts[0]) + " " + ids[meio]);

            if(parts[1].equals(nome)){
                resp = true;
                esq = dir + 1;
            } else if (getId(nome) > ids[meio]) {
                esq = meio + 1;
            } else {
                dir = meio - 1;
            }
            arq.close();
        }
        return resp;
    }


    public static void ordenar(){
        quicksort(0, n-1);
    }

    public static void swap(int i, int j) {
      int temp = ids[i];
      ids[i] = ids[j];
      ids[j] = temp;
   }

    public static void quicksort(int esq, int dir) {
        int i = esq, j = dir;
        int pivo = ids[(dir+esq)/2];
        while (i <= j) {
            while (ids[i] < pivo) i++;
            while (ids[j] > pivo) j--;
            if (i <= j) {
                swap(i, j);
                i++;
                j--;
            }
        }
        if (esq < j)  quicksort(esq, j);
        if (i < dir)  quicksort(i, dir);
    }


    public static void main(String[] args){
        String id = "";
        String nome = "";
        int i = 0;
        int j = 0;
        double inicio, fim;

        
        do{
            id = myIo.readLine();
            if (!id.equalsIgnoreCase("FIM")){
                ids[i] = Integer.parseInt(id);
                i++;
                n++;
            }
        } while (!id.equalsIgnoreCase("FIM"));

        /*for(int k = 0; k < n; k++){
            myIo.print(ids[k] + " ");
        }*/
        //myIo.println(" ");
        ordenar();
       
        /*for(int l = 0; l < n; l++){
            myIo.print(ids[l] + " ");
        }*/

        //inicio = System.currentTimeMillis();
        do{
            nome = myIo.readLine();
            //myIo.println("nome: " + nome);
            if (!nome.equalsIgnoreCase("FIM")){
                myIo.println(pesquisar(nome)?"SIM":"NAO");
            }
        } while (!nome.equalsIgnoreCase("FIM"));
        //fim = System.currentTimeMillis();
        
        //String tempo = (fim-inicio)/1000 + "\t";
        //String conteudo = "816479\t" + tempo + comp;
        //arq.openWriteClose("matrícula_sequencial.txt", conteudo);
    }
}

File: AEDSII/TP02/q04/q04.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int compNome(int id1, int id2){
    //printf("comp");
    char *nome1 = malloc (120);
    char *nome2 = malloc (120);
    FILE *arq1;
    FILE *arq2;

    arq1 = fopen("/tmp/players.csv", "r");
    for (int i = 0; i < id1 + 2; i++){
        fgets(nome1, 120, arq1);
    }
    fclose(arq1);
    if ((strlen(nome1) > 0) && (nome1[strlen (nome1) - 1] == '\n')){
        nome1[strlen (nome1) - 1] = '\0';
    }

    arq2 = fopen("/tmp/players.csv", "r");
    for (int i = 0; i < id2 + 2; i++){
        fgets(nome2, 120, arq2);
    }
    fclose(arq2);
    if ((strlen(nome2) > 0) && (nome2[strlen (nome2) - 1] == '\n')){
        nome2[strlen (nome2) - 1] = '\0';
    }

    //printf("%s\n", nome1);
    //printf("%s\n", nome2);
    
    int j = 0;
    char *info1 = strtok(nome1, ",");
    char *teste1 = malloc(40);
    char *teste2 = malloc(40);
    int result = -1;

    while (j < 2)
        {
            if(j == 1){
                teste1 = info1;
            }
            info1 = strtok(NULL, ",");
            j++;
        }
        j = 0;

        char *info2 = strtok(nome2, ",");
        while (j < 2)
        {
            if(j == 1){
                teste2 = info2;
            }
            info2 = strtok(NULL, ",");
            j++;
        }

        if(strcmp(teste1, teste2) > 0){
            result = 1;
        } else if (strcmp(teste1, teste2) == 0){
            result = 0;
        }

        //printf("%s %s\n", teste1, teste2);
    return result;
}

int compNome2(int id1, char* nome){
    //printf("comp");
    char *nome1 = malloc (120);
    FILE *arq1;
    

    arq1 = fopen("/tmp/players.csv", "r");
    for (int i = 0; i < id1 + 2; i++){
        fgets(nome1, 120, arq1);
    }
    //printf("leu");
    fclose(arq1);
    if ((strlen(nome1) > 0) && (nome1[strlen (nome1) - 1] == '\n')){
        nome1[strlen (nome1) - 1] = '\0';
    }

    //printf("leu2");

    
    int j = 0;
    char *info1 = strtok(nome1, ",");
    char *teste1 = malloc(40);
    int result = -1;

    while (j < 2)
        {
            if(j == 1){
                teste1 = info1;
            }
            info1 = strtok(NULL, ",");
            j++;
        }
        j = 0;


        if(strcmp(teste1, nome) > 0){
            result = 1;
        } else if (strcmp(teste1, nome) == 0){
            result = 0;
        }

        //printf("%s %s\n", teste1, nome);
        free(nome1);
    return result;
}

void swap(int *i, int *j) {
    int temp = *i;
    *i = *j;
    *j = temp;
}

void ordenar(int *array, int n){
    for (int i = 0; i < (n - 1); i++) {
      int menor = i;
      for (int j = (i + 1); j < n; j++){
         if (compNome(array[menor], array[j]) > 0){
            menor = j;
         }
      }
      swap(&array[menor], &array[i]);
   }
}

int pesquisar(int *id, char *nome, int inicio, int fim){
    //printf("pesq");
    int i = (inicio + fim) / 2;

    if (inicio > fim) {  // Ponto de parada
        //printf("n achou");
        return -1;
    }

    if (compNome2(id[i],nome) == 0) {  // Item encontrado 
        //printf("achou");
        return 1;
    }

    if (compNome2(id[i], nome) < 0) {  // Item está à direita 
        //printf("maior");
        return pesquisar(id, nome, i+1 , fim );
    
    } else {  //Item está à esquerda 
        //printf("menor");
        return pesquisar(id, nome, inicio, i-1);
    }
}


int main (void){
    int ids[80];
    char nome[40];
    char lixo[10];
    char *resp = malloc(10);
    int id = 0;
    int lastId = -1;
    int i = 0;
    int j = 0;

   while (id != lastId){
        lastId = id;
        scanf("%d", &id);
        if (id != lastId){
            ids[i] = id;
            i++;
        }
    }

   ordenar(ids, i);

   /*for(j = 0; j < i; j++){
    printf("%d", ids[j]);
   }*/

    //printf("teste1");
    fgets(lixo, 10, stdin);
    //printf("%s", lixo);

    do{
        //printf("teste2");
        fgets(nome, 40, stdin);
        if ((strlen(nome) > 0) && (nome[strlen (nome) - 1] == '\n')){       // formatacao da entrada
            nome[strlen (nome) - 1] = '\0';
        }
        
               
        if(strcmp(nome, "FIM\0")!=0){
            
            //printf("%s", nome);
            
            resp = (pesquisar(ids, nome, 0, i-1) > 0) ? "SIM\n" : "NAO\n";
            printf("%s", resp);
        }
    } while (strcmp(nome, "FIM\0")!=0); 
}